Explorer/Assets/DCL/Web3Authentication/DappWeb3Authenticator.cs

```
@@ -0,0 +1,214 @@
+        private const int TIMEOUT_SECONDS = 30;
+
+        private readonly IWebBrowser webBrowser;
+
+        private SocketIO? webSocket;
+        private UniTaskCompletionSource<DappSignatureResponse>? signatureOutcomeTask;
+
+        public IWeb3Identity? Identity { get; private set; }
+
+        public DappWeb3Authenticator(IWebBrowser webBrowser)
+        {
+            this.webBrowser = webBrowser;
+        }
+
+        public void Dispose()
+        {
+            webSocket?.Dispose();
+            Identity?.Dispose();
+        }
+
+        /// <summary>
+        ///     1. An authentication request is sent to the server
+        ///     2. Open a tab to let the user sign through the browser with his custom installed wallet
+        ///     3. Use the signature information to generate the identity
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        /// <exception cref="Web3SignatureException"></exception>
+        public async UniTask<IWeb3Identity> LoginAsync(CancellationToken cancellationToken)
+        {
+            try
+            {
+                InitializeWebSocket();
+
+                await ConnectToServerAsync();
+
+                var ephemeralAccount = NethereumAccount.CreateRandom();
+
+                // 1 week expiration day, just like unity-renderer
+                DateTime expiration = DateTime.UtcNow.AddDays(7);
+                string ephemeralMessage = CreateEphemeralMessage(ephemeralAccount, expiration);
+
+                SignatureIdResponse authenticationResponse = await RequestAuthenticationAsync(ephemeralMessage, cancellationToken);
+
+                await UniTask.SwitchToMainThread(cancellationToken);
+
+                LetUserSignThroughDapp(authenticationResponse.requestId);
+                DappSignatureResponse signature = await WaitForSignatureOutcome(cancellationToken);
```

I belive somewhere here error handling should exist. Otherwise only a happy path is handled

---

What kind of error you think we could handle before breaking the operation process? All the steps are necessary to get a valid authentication, and if any fails, perhaps a login retry should be done by the caller (outside this class), but we dont have a real login scenario yet

---

Also, the server does not emit any kind of special error in the process, it just follows the happy path

---

The only handling im doing at the moment is the signature expiration

---

There is `code` in the response: I believe it can indicate the error. Please sync with dApps team. I am highlighting that you can't continue the normal flow if `SignatureIDResponse` has failed: you should throw an exception immediately. And handling that exception will be the responsibility of the upper-level flow.

---

The `code` is the verification code which should be the same on the webapp. The user manually verifies this number as an extra security step. I added validations in the responses nonetheless

---

Thanks for adding validation 💪 It will cover our needs.
I'd like to highlight though that error reporting should have been done on the server side. I am not entirely sure why it's not done there as it's a standard flow for any kind of communication including authentication such as Facebook, Google Play Services, etc.

E.g. for GPS:
```
using GooglePlayGames;

    public void Start() {
      PlayGamesPlatform.Instance.Authenticate(ProcessAuthentication);
    }

    internal void ProcessAuthentication(SignInStatus status) {
      if (status == SignInStatus.Success) {
        // Continue with Play Games Services
      } else {
        // Disable your integration with Play Games Services or show a login button
        // to ask users to sign-in. Clicking it should call
        // PlayGamesPlatform.Instance.ManuallyAuthenticate(ProcessAuthentication).
      }
    }
}
```